Alignment/OfflineValidation/plugins/OverlapValidation.cc

```diff
@@ -0,0 +1,876 @@
+{
+  using namespace edm;
+  //
+  // mag field & search tracker
+  //
+  edm::ESHandle<MagneticField> magFieldHandle;
+  iSetup.get<IdealMagneticFieldRecord>().get(magFieldHandle);
+  magField_ = magFieldHandle.product();
+  //
+  // propagator
+  //
+  AnalyticalPropagator propagator(magField_,anyDirection);
+  //
+  // geometry
+  
+
+  //
+  edm::ESHandle<TrackerGeometry> geometryHandle;
+  iSetup.get<TrackerDigiGeometryRecord>().get(geometryHandle);
+  trackerGeometry_ = geometryHandle.product();
+  //
+  // make associator for SimHits
+  //
+  TrackerHitAssociator* associator;
+  if(doSimHit_) {
+    TrackerHitAssociator::Config hitassociatorconfig(config_, consumesCollector());
+    associator = new TrackerHitAssociator(iEvent, hitassociatorconfig);
+  } else {
+    associator = nullptr;
+  
+
+  }  
+
+
+
+
+
+
+
+
+//if(doSimHit_) associator = new TrackerHitAssociator(iEvent, config_); else associator = 0; 
+
+  //
+  // trajectories (from refit)
+  //
+  //typedef vector<Trajectory> TrajectoryCollection;
+  edm::Handle<TrajectoryCollection> trajectoryCollectionHandle;
+  iEvent.getByToken(trajectoryToken_,trajectoryCollectionHandle);
+  const TrajectoryCollection* trajectoryCollection = trajectoryCollectionHandle.product();
+
+  //
+  // loop over trajectories from refit
+ edm::ESHandle<TrackerTopology> tTopoHandle;
+ iSetup.get<TrackerTopologyRcd>().get(tTopoHandle);
+ const TrackerTopology* const tTopo = tTopoHandle.product(); 
+ for ( TrajectoryCollection::const_iterator it=trajectoryCollection->begin();
+	it!=trajectoryCollection->end(); ++it )  analyze(*it,propagator,*associator, tTopo);
+  
+  run_ = iEvent.id().run();
+  event_ = iEvent.id().event();
+  
+
+  
+}
+
+
+void
+OverlapValidation::analyze (const Trajectory& trajectory,
+		     const Propagator& propagator,
+		     TrackerHitAssociator& associator,  const TrackerTopology* const tTopo)
+{
+  typedef std::pair<const TrajectoryMeasurement*, const TrajectoryMeasurement*> Overlap;
+  typedef vector<Overlap> OverlapContainer;
+  ++overlapCounts_[0];
+
+  OverlapContainer overlapHits;
+
+  // quality cuts on trajectory
+  // min. # hits / matched hits
+
+  if ( trajectory.foundHits()<6 )  return;
+  if ( ChiSquaredProbability((double)( trajectory.chiSquared() ),(double)( trajectory.ndof(false) )) < 0.001 ) return;
+  //
+  // loop over measurements in the trajectory and calculate residuals
+  //
+  
+
+  vector<TrajectoryMeasurement> measurements(trajectory.measurements());
+  for ( vector<TrajectoryMeasurement>::const_iterator itm=measurements.begin();
+	itm!=measurements.end(); ++itm ) {
+    //
+    // skip "invalid" (i.e. missing) hits
+    //
+    ConstRecHitPointer hit = itm->recHit();
+    DetId id = hit->geographicalId();
+    int layer(layerFromId(id, tTopo));
+    int subDet = id.subdetId();
+
+    if ( !hit->isValid() ) {
+      edm::LogVerbatim("OverlapValidation")  << "Invalid";
+      continue;
+    }
+    if (barrelOnly_ && (subDet==4 || subDet==6) ) return;
```

Comment:
please use `StripSubdetector::TID` and `StripSubdetector::TEC` to improve readability.

Reply:
Done

Reply:
Done